js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts

```diff
@@ -0,0 +1,591 @@
+type FieldDataMapping<O extends FieldsMapping> = {
+    readonly [K in keyof O]: O[K]['__fieldValueType'];
+};
+
+// Extract mapping of field name to it's parsable data type. For most fields
+// this is the same as the underlying type but could be something different, eg.
+// parsing a string => number
+type FieldDataMappingRaw<O extends FieldsMapping> = {
+    [K in keyof O]?: O[K]['__parsableValueType'];
+};
+
+export type SinglePrimaryKey = string | number;
+export type CompoundPrimaryKey = { [fieldName: string]: SinglePrimaryKey };
+export type PrimaryKey = SinglePrimaryKey | CompoundPrimaryKey;
+
+/**
+ * Creates a ViewModel class with the specified fields.
+ *
+ * ```js
+ * const fields = {
+ *     userId: new IntegerField({ label: 'User ID' })
+ *     firstName: new CharField({ label: 'First Name' }),
+ *     // label is optional; will be generated as 'Last name'
+ *     lastName: new CharField(),
+ * };
+ * // Options are all optional
+ * const options = {
+ *     // If not specified will create a default field called 'id'
+ *     pkFieldName: 'userId',
+ * };
+ * class User extends ViewModel(fields, options) {
+ *     // Optional; default cache is usually sufficient
+ *     static cache = new MyCustomCache();
+ *
+ *     // Used to describe a single user
+ *     static label = 'User';
+ *     // User to describe an indeterminate number of users
+ *     static labelPlural = 'Users';
+ * }
+ * ```
+ */
+type ViewModelInterface<
+    FieldMappingType extends FieldsMapping,
+    InstanceFieldMappingType extends FieldMappingType,
+    PkFieldType extends string | string[] = string | string[],
+    PkType extends PrimaryKey = PrimaryKey
+> = FieldDataMapping<FieldMappingType> & {
+    /**
+     * Get the actual ViewModel class for this instance
+     */
+    _model: ViewModelConstructor<FieldMappingType, PkFieldType, PkType>;
+    /**
+     * Return the data for this record as an object
+     */
+    toJS(): () => FieldDataMapping<FieldMappingType>;
+    /**
+     * Compares two records to see if they are equivalent.
+     *
+     * - If the ViewModel is different then the records are always considered different
+     * - If the records were initialised with a different set of fields then they are
+     *   considered different even if the common fields are the same and other fields are
+     *   all null
+     */
+    isEqual(record: ViewModelInterface<FieldMappingType, InstanceFieldMappingType>): boolean;
+    /**
+     * Clone this record, optionally with only a subset of the fields
+     */
+    clone(
+        fieldNames?: string[]
+    ): ViewModelInterface<FieldMappingType, InstanceFieldMappingType, PkFieldType, PkType>;
+    /**
+     * Returns the primary key value(s) for this instance. This is a shortcut
+     * for using the primary key field name(s) directly.
+     */
+    readonly _pk: PkType;
+    readonly _data: FieldDataMapping<InstanceFieldMappingType>;
+    readonly _assignedFields: (keyof InstanceFieldMappingType)[];
```

Comment:
What is `_data` and `_assignedFields`? (I may be misunderstanding these, may affect some of my later comments)

Reply:
`_data` is the data for this object. shortcut for the assigned data

`_assignedFields` is the fields set on this object (eg. partial record with these fields set)